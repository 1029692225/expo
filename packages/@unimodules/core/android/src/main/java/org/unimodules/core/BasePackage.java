package org.unimodules.core;

import android.content.Context;

import org.unimodules.core.interfaces.InternalModule;
import org.unimodules.core.interfaces.Package;
import org.unimodules.core.interfaces.SingletonModule;

import java.util.Collections;
import java.util.List;

// This class should not be used. Implement org.unimodules.core.interfaces.Package instead of extending this class
// Remove once no one extends it.
public class BasePackage implements Package {
  @Override
  public List<? extends InternalModule> createInternalModules(Context context) {
    return Collections.emptyList();
  }

  @Override
  public List<? extends ExportedModule> createExportedModules(Context context) {
    return Collections.emptyList();
  }

  @Override
  public List<? extends ViewManager> createViewManagers(Context context) {
    return Collections.emptyList();
  }

  @Override
  public List<? extends SingletonModule> createSingletonModules(Context context) {
    return Collections.emptyList();
  }
}
