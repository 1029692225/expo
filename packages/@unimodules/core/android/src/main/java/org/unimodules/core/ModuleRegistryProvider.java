package org.unimodules.core;

import android.content.Context;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import org.unimodules.core.interfaces.InternalModule;
import org.unimodules.core.interfaces.Package;
import org.unimodules.core.interfaces.SingletonModule;

/**
 * Builder for {@link ModuleRegistry}. Override this class to add some custom
 * modules from outside of {@link Package} ecosystem.
 */
public class ModuleRegistryProvider {
  private List<Package> mPackages;
  protected String mAppId;

  public ModuleRegistryProvider(List<Package> initialPackages, String appId) {
    mPackages = initialPackages;
    mAppId = appId;
  }

  protected List<Package> getPackages() {
    return mPackages;
  }

  public ModuleRegistry get(Context context) {
    return new ModuleRegistry(mAppId,
            createInternalModules(context),
            createExportedModules(context),
            createViewManagers(context),
            createSingletonModules(context)
    );
  }

  protected Collection<InternalModule> createInternalModules(Context context) {
    Collection<InternalModule> internalModules = new ArrayList<>();
    for (Package pkg : getPackages()) {
      internalModules.addAll(pkg.createInternalModules(context));
    }
    return internalModules;
  }

  protected Collection<ExportedModule> createExportedModules(Context context) {
    Collection<ExportedModule> exportedModules = new ArrayList<>();
    for (Package pkg : getPackages()) {
      exportedModules.addAll(pkg.createExportedModules(context));
    }
    return exportedModules;
  }

  protected Collection<ViewManager> createViewManagers(Context context) {
    Collection<ViewManager> viewManagers = new ArrayList<>();
    for (Package pkg : getPackages()) {
      viewManagers.addAll(pkg.createViewManagers(context));
    }
    return viewManagers;
  }

  public Collection<SingletonModule> createSingletonModules(Context context) {
    Collection<SingletonModule> singletonModules = new ArrayList<>();
    for (Package pkg : getPackages()) {
      singletonModules.addAll(pkg.createSingletonModules(context));
    }
    return singletonModules;
  }
}
