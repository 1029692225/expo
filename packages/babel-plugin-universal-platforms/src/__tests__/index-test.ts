import { transformSync } from '@babel/core';
// @ts-ignore
import dedent from 'dedent';
import universalPlatform, { Options } from '..';

function transform(input: string, options: Options): string {
  const value = transformSync(dedent(input), {
    plugins: [[universalPlatform, options]],
    babelrc: false,
    configFile: false,
  });

  return value == null ? '' : (value.code as string);
}

it(`removes __DEV__ and process.env.NODE_ENV`, () => {
  const code = transform(
    `
  if (__DEV__) {
    console.log("DEV MODE")
  } else {
    console.log("PROD MODE")
  }
  if (process.env.NODE_ENV === 'development') {
    console.log("DEV MODE")
  } else {
    console.log("PROD MODE")
  }
  `,
    {
      platform: `ios`,
      mode: 'development',
    }
  );

  expect(code).toMatch('DEV');
  expect(code).not.toMatch('PROD');
  expect(code).toMatchSnapshot();
});

// react-native-web redefines this value.
// Terser should shake it during the bundling.
it(`keeps __DEV__ redefinition`, () => {
  const code = transform(
    `
  const __DEV__ = process.env.NODE_ENV !== 'production';
  `,
    {
      platform: `web`,
      mode: 'development',
    }
  );

  expect(code).toMatch('__DEV__');
  expect(code).not.toMatch('process.env.NODE_ENV');
});

// react-native-web redefines this value.
// Terser should shake it during the bundling.
it(`should pass over process.env.NODE_ENV redefinition`, () => {
  const code = transform(
    `
  process.env.NODE_ENV = 'production';
  `,
    {
      platform: `web`,
      mode: 'development',
    }
  );
  expect(code).toMatch(`process.env.NODE_ENV = 'production'`);
});

it(`switch statement is converted for Terser`, () => {
  const code = transform(
    `
    switch (Platform.OS) {
      case 'web':
        console.log('web');
        break;
      default:
        console.log('default');
        break;
    }
  `,
    {
      platform: `web`,
      mode: 'development',
    }
  );
  expect(code).toMatch('switch ("web")');
});

describe('if Statements', () => {
  const DEFAULT_BLOCK = `
  if (Platform.OS === 'ios') {
    console.log('iOS')
  } else if (Platform.OS == "android") {
    console.log('Android')
  } else if ("web" === Platform.OS) {
    console.log('web')
  } else {
    console.log('and beyond')
  }
  `;

  for (const platform of ['iOS', 'Android', 'web', 'custom']) {
    it(`only saves ${platform} code`, () => {
      const code = transform(DEFAULT_BLOCK, {
        platform: platform.toLowerCase(),
        mode: 'development',
      });

      expect(code).not.toMatch('Platform.OS');
      expect(code).toMatchSnapshot();
    });
  }
});

describe('Platform.OS', () => {
  it(`is converted to a static value for Terser`, () => {
    const code = transform(
      `
    const value = Platform.OS;
    `,
      {
        platform: `ios`,
        mode: 'development',
      }
    );

    expect(code).not.toMatch('Platform.OS');
    expect(code).toMatchSnapshot();
  });
});

describe('Platform.select', () => {
  it('should replace Platform.select with value for default case if there is no matching platform', () => {
    const code = transform(
      `
      const value = Platform.select({
        android: 'android',
        ios: () => {},
        default: { test: true },
        web: 'what?'
      });
      `,
      {
        platform: 'custom',
        mode: 'development',
      }
    );

    expect(code).toMatch(/const value = {\s*test: true\s*}/);
    expect(code).not.toMatch('Platform.select');
    expect(code).toMatchSnapshot();
  });

  it('should replace Platform.select with value for matching platform', () => {
    const code = transform(
      `
      const defaults = { default: 'default' };
      const value = Platform.select({
        ios: () => {},
        ...defaults,
      });
      `,
      {
        platform: 'ios',
        mode: 'development',
      }
    );

    expect(code).toMatch('const value = () => {};');
    expect(code).not.toMatch('Platform.select');
    expect(code).toMatchSnapshot();
  });

  it('should remove non-matching platforms but leave Platform.select', () => {
    const code = transform(
      `
      const defaults = { default: 'default' };
      const value = Platform.select({
        ios: () => {},
        ...defaults,
      });
      `,
      {
        platform: 'android',
        mode: 'development',
      }
    );

    expect(code).toMatch('...defaults');
    expect(code).toMatch('Platform.select');
    expect(code).not.toMatch('ios');
    expect(code).toMatchSnapshot();
  });

  it('should replace Platform.select with undefined if no cases are matching', () => {
    const code = transform(
      `
      const value = Platform.select({
        android: () => {},
      });
      `,
      {
        platform: 'ios',
        mode: 'development',
      }
    );

    expect(code).toEqual('const value = undefined;');
  });

  it('should remove non-matching platforms but leave Platform.select with unknown cases 1', () => {
    const code = transform(
      `
      const value = Platform.select({
        ios: false,
        ['andr' + 'oid']: () => {},
      });
      `,
      {
        platform: 'android',
        mode: 'development',
      }
    );

    expect(code).toMatch("['andr' + 'oid']: () => {}");
    expect(code).toMatch('Platform.select');
    expect(code).not.toMatch('ios');
    expect(code).toMatchSnapshot();
  });

  it('should remove non-matching platforms but leave Platform.select with unknown cases and default case', () => {
    const code = transform(
      `
      const value = Platform.select({
        default: 'default',
        ['w' + 'eb']() {}
      });
      `,
      { platform: 'web', mode: 'development' }
    );

    expect(code).toMatch("['w' + 'eb']() {}");
    expect(code).toMatch('Platform.select');
    expect(code).toMatch("default: 'default',");
    expect(code).toMatchSnapshot();
  });

  it('should remove non-matching platforms but leave Platform.select with method', () => {
    const code = transform(
      `
      const value = Platform.select({
        ios: false,
        android() {},
        web: false,
      });
      `,
      {
        platform: 'android',
        mode: 'development',
      }
    );

    expect(code).toMatch('android() {}');
    expect(code).toMatch('Platform.select');
    expect(code).not.toMatch('ios');
    expect(code).not.toMatch('web');
    expect(code).toMatchSnapshot();
  });
});
