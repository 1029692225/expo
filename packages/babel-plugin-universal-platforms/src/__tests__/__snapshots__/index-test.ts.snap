// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Platform.OS is converted to a static value for Terser 1`] = `"const value = \\"ios\\";"`;

exports[`Platform.select should remove non-matching platforms but leave Platform.select 1`] = `
"const defaults = {
  default: 'default'
};
const value = Platform.select({ ...defaults
});"
`;

exports[`Platform.select should remove non-matching platforms but leave Platform.select with method 1`] = `
"const value = Platform.select({
  android() {}

});"
`;

exports[`Platform.select should remove non-matching platforms but leave Platform.select with unknown cases 1 1`] = `
"const value = Platform.select({
  ['andr' + 'oid']: () => {}
});"
`;

exports[`Platform.select should remove non-matching platforms but leave Platform.select with unknown cases and default case 1`] = `
"const value = Platform.select({
  default: 'default',

  ['w' + 'eb']() {}

});"
`;

exports[`Platform.select should replace Platform.select with value for default case if there is no matching platform 1`] = `
"const value = {
  test: true
};"
`;

exports[`Platform.select should replace Platform.select with value for matching platform 1`] = `
"const defaults = {
  default: 'default'
};

const value = () => {};"
`;

exports[`if Statements only saves Android code 1`] = `
"{
  console.log('Android');
}"
`;

exports[`if Statements only saves custom code 1`] = `
"{
  console.log('and beyond');
}"
`;

exports[`if Statements only saves iOS code 1`] = `
"{
  console.log('iOS');
}"
`;

exports[`if Statements only saves web code 1`] = `
"{
  console.log('web');
}"
`;

exports[`removes __DEV__ and process.env.NODE_ENV 1`] = `
"{
  console.log(\\"DEV MODE\\");
}
{
  console.log(\\"DEV MODE\\");
}"
`;
